o/fdestate: add keyslot change conflict detection

ZeyadYasser opened the pull request:
This PR adds keyslot change conflict detection to prevent two changes affecting the same key slot from being started in parallel.

This is part 1 of implementing handlers for recovery key replacement which I had to break down as it was getting quite large. for context, full e2e implementation can be found in https://github.com/canonical/snapd/pull/15587.

pedronis commented on overlord/fdestate/conflict.go:
```diff
@@ -0,0 +1,78 @@
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+type ChangeConflictError struct {
```
@bboozzoo is correct, we will translation code in daemon to produce client errors with the appropriate status (409) and kind and extra information

pedronis commented:
something to keep in mind, when we start dealing with removing any keyslot (not just recovery key ones) or adding new platform keyslots we also need to deal with conflicts with resealing operations

ernestl reviewed (commented):
Wondering about the need/benefit for parallel keyslot changes. Perhaps better to limit to one keyslot change at a time at API level.

Simplicity of one-at-time seems welcome given the existing challenge around idempotency & destructive changes. Expect an API user requesting multiple replacements would expect all or nothing.

This discussion also seems relevant, perhaps worth revisiting: https://chat.canonical.com/canonical/pl/oppimawttif7mdentzzsnwzd6o

ZeyadYasser commented:
> Wondering about the need/benefit for parallel keyslot changes. Perhaps better to limit to one keyslot change at a time at API level.
> 
> Simplicity of one-at-time seems welcome given the existing challenge around idempotency & destructive changes. Expect an API user requesting multiple replacements would expect all or nothing.
> 
> This discussion also seems relevant, perhaps worth revisiting: https://chat.canonical.com/canonical/pl/oppimawttif7mdentzzsnwzd6o

Thank you for the suggestion, I updated the PR to do more generic conflict detection for all FDE changes (including dbx updates).

Also, implemented suggestion for blocking key slot operations from running in parallel for safety.

pedronis commented on overlord/fdestate/conflict.go:
```diff
@@ -0,0 +1,47 @@
+				ChangeKind: chg.Kind(),
+				ChangeID:   chg.ID(),
+			}
+		case "fde-replace-recovery-key":
```
shouldn't we check instead if any of the task in the change have keyslots attached?

ZeyadYasser replied:
updated to check for tasks with attached keyslots as a fallback.

pedronis reviewed (commented):
question

pedronis commented on daemon/api_system_volumes.go:
```diff
@@ -141,7 +142,7 @@ func postSystemVolumesActionReplaceRecoveryKey(c *Command, req *systemVolumesAct
 
-	chg := st.NewChange("replace-recovery-key", "Replace recovery key")
+	chg := newChange(st, "fde-replace-recovery-key", "Replace recovery key", []*state.TaskSet{ts}, nil)
 	chg.AddAll(ts)
```
we don't need this anymore, if you pass it above

ZeyadYasser replied:
thanks for catching this, my bad. updated the unit test to catch this bug.

pedronis reviewed (changes requested):
the change in daemon doesn't look completely correct

ernestl commented on overlord/fdestate/fdemgr.go:
```diff
@@ -404,6 +420,7 @@ func (m *FDEManager) GetKeyslots(keyslotRefs []KeyslotRef) (keyslots []Keyslot,
 		keyslots = append(keyslots, matchedContainerKeyslots...)
 	}
 
+	// XXX: return error if len(keyslots) != keyslotRefs to indicate duplicates?
```
when will this happen?

ernestl replied:
+1 for strict

pedronis commented on daemon/api_system_volumes.go:
```diff
@@ -141,8 +142,7 @@ func postSystemVolumesActionReplaceRecoveryKey(c *Command, req *systemVolumesAct
 		return BadRequest("cannot replace recovery key: %v", err)
```
this code needs to do error conversion correctly to produce a 409 conflict error if there's a conflict, we also need a test for that

ZeyadYasser commented on overlord/fdestate/dbx.go:
great catch, thank you!

ernestl commented on overlord/fdestate/errors_test.go:
```diff
@@ -0,0 +1,38 @@
+		{ContainerRole: "some-role", Name: "some-slot-1"},
+		{ContainerRole: "some-role", Name: "some-slot-2"},
+	}}
+	c.Check(err.Error(), Equals, `key slot references [(container-role: "some-role", name: "some-slot-1"), (container-role: "some-role", name: "some-slot-2")] not found`)
```
[{...}, {...}]

pedronis commented on overlord/fdestate/fdemgr.go:
```diff
@@ -128,6 +129,17 @@ func Manager(st *state.State, runner *state.TaskRunner) (*FDEManager, error) {
 	runner.AddHandler("remove-keys", m.doRemoveKeys, nil)
 	runner.AddHandler("rename-keys", m.doRenameKeys, nil)
+	runner.AddBlocked(func(t *state.Task, running []*state.Task) bool {
+		if strings.HasPrefix(t.Change().Kind(), "fde-") || t.Has("keyslots") {
```
is the change kind thing an optimization, I wouldn't do it, it seems fragile, one point of tasks is to be able to mix them, what we can do as I proposed is to add a fde- prefix to the task names instead

ZeyadYasser replied:
not intended as an optimization but to match `checkFDEChangeConflict` change prefix check suggested here https://github.com/canonical/snapd/pull/15588#discussion_r2166358610, I can switch both the `AddBlocked` and `checkFDEChangeConflict` to a task level check instead and add `fde-` prefix to task names.

@bboozzoo @pedronis, would this mean also pre-fixing dbx tasks with `fde-` or leave it as it?

pedronis reviewed (commented):
I'm a bit confused by the AddBlocked changes

pedronis commented on overlord/fdestate/dbx.go:
```diff
@@ -570,7 +570,12 @@ func checkDBXChangeConflicts(st *state.State) error {
+	}
+
+	// make sure no changes related to FDE are in progress
+	return checkFDEChangeConflict(st)
```
yea, moving this check one level up seems fine to me too

bboozzoo commented on overlord/fdestate/conflict.go:
```diff
@@ -0,0 +1,63 @@
+			// try to catch changes/tasks that could have been missed
+			// and log a warning.
+			for _, t := range chg.Tasks() {
+				if strings.HasPrefix(t.Kind(), "fde-") {
```
if you're feeling fancy this could be made into a helper `fund isFDETask(t *state.Task) bool` at some point.

bboozzoo reviewed (approved):
LGTM
